### PR TITLE
Fix errata in Kubernetes controllers

### LINKED ISSUE #6493: Radius controller updates deployment spec when Radius is enabled
### Steps to reproduce (required)

Follow the Kubernetes & Helm tutorial up to step 4.

### Observed behavior (required)

In step 4 Radius updates the deployment to inject a secret reference:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "2"
    meta.helm.sh/release-name: demo
    meta.helm.sh/release-namespace: demo
    radapp.io/configuration-hash: bf21a9e8fbc5a3846fb05b4fa0859e0917b2202f
    radapp.io/enabled: "true"
    radapp.io/status: '{"scope":"/planes/radius/local/resourceGroups/default-demo","application":"/planes/radius/local/resourceGroups/default-demo/providers/Applications.Core/applications/demo","environment":"/planes/radius/local/resourcegroups/default/providers/Applications.Core/environments/default","container":"/planes/radius/local/resourceGroups/default-demo/providers/Applications.Core/containers/webapp","phrase":"Ready"}'
  creationTimestamp: "2023-10-17T18:02:30Z"
  finalizers:
  - radapp.io/deployment-finalizer
  generation: 4
  labels:
    app.kubernetes.io/managed-by: Helm
  name: webapp
  namespace: demo
  resourceVersion: "4282"
  uid: 0f87f944-eff1-4cd4-a546-da5a6eee95ca
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: webapp
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: webapp
    spec:
      containers:
      - env:
        - name: CONNECTION_REDIS_URL
          valueFrom:
            secretKeyRef:
              key: url
              name: redis-secret
        envFrom:
        - secretRef:
            name: webapp-connections
            optional: false
        image: radius.azurecr.io/tutorial/webapp:edge
        imagePullPolicy: IfNotPresent
        name: webapp
        ports:
        - containerPort: 3000
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2023-10-17T18:03:15Z"
    lastUpdateTime: "2023-10-17T18:03:15Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2023-10-17T18:02:30Z"
    lastUpdateTime: "2023-10-17T18:07:07Z"
    message: ReplicaSet "webapp-54c668f4bf" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 4
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1

```

### Desired behavior (required)

Radius does not update the deployment unless the user declares a connection.

### Workaround (optional)

This is non-blocking but goes against one of our value propositions. We want to tell users that enabling Radius doesn't change any behaviors. This **does** change a behavior.

[AB#9832](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/9832)

### PR DESCRIPTION
**Pull request by rynowak**
# Description

This change fixes the minor issues I encountered walking through the tutorial.

There are three issues:

- Radius injects a connection secret even when there are no connections (#6493)
- Radius does not rollout new pods when connections change
- `rad app graph` doesn't deal well with mixed casing

## Type of change

- This pull request fixes a bug in Radius and has an approved issue (issue link required).

Fixes: #6493

### Summary
:memo::recycle::hammer:

This pull request improves the handling of secrets for deployments in the `reconciler` package and adds a `kubernetes` package with a utility function to hash secret data. It also fixes some casing issues in the tests and documents a workaround for a `rad app graph` bug.

> _Sing, O Muse, of the skillful coder who changed the reconciler_
> _To handle secrets with care and grace, avoiding duplication and error_
> _He moved the `hashSecretData` function to the `kubernetes` package_
> _Like Zeus who moved the thunderbolt to his mighty hand from the forge of Hephaestus_

### Walkthrough
*  Move the `hashSecretData` function from the `container` package to the `kubernetes` package and make it public as `HashSecretData` ([link](https://github.com/radius-project/radius/pull/6494/files?diff=unified&w=0#diff-3da981bf20df07918a646eafcd867d6c95abac1f3aef1ed1b50a29f3f4cc8636L21), [link](https://github.com/radius-project/radius/pull/6494/files?diff=unified&w=0#diff-3da981bf20df07918a646eafcd867d6c95abac1f3aef1ed1b50a29f3f4cc8636L656-R655), [link](https://github.com/radius-project/radius/pull/6494/files?diff=unified&w=0#diff-3da981bf20df07918a646eafcd867d6c95abac1f3aef1ed1b50a29f3f4cc8636L845-L864), [link](https://github.com/radius-project/radius/pull/6494/files?diff=unified&w=0#diff-5d6c3370ac092d779d7f74cb9e6371c114df3d684a99870dd58f288f02026a9aR1-R47)). This is done to provide a common utility function to hash the data in a secret and use it as a Kubernetes annotation to force a deployment to redeploy pods when the secret changes.

### CODE REVIEW

File: pkg/controller/reconciler/deployment_reconciler.go [hunk 1]
```diff
@@ -487,6 +488,19 @@ func (r *DeploymentReconciler) updateDeployment(ctx context.Context, deployment
 	// We store the connection values in a Kubernetes secret and then use the secret to populate environment variables.
 	secretName := client.ObjectKey{Namespace: deployment.Namespace, Name: fmt.Sprintf("%s-connections", deployment.Name)}
 
+	if len(annotations.Configuration.Connections) == 0 {
```
Comment (rynowak) [inline pkg/controller/reconciler/deployment_reconciler.go:491]: This handles #6493 - we need to keep things clean when there are no connections defined.

File: pkg/controller/reconciler/deployment_reconciler.go [hunk 2]
```diff
@@ -546,6 +560,14 @@ func (r *DeploymentReconciler) updateDeployment(ctx context.Context, deployment
 		}
 	}
 
+	// Add the hash of the secret data to the Pod definition. This will force a rollout when the secrets
+	// change.
+	hash := kubernetes.HashSecretData(secret.Data)
+	if deployment.Spec.Template.ObjectMeta.Annotations == nil {
+		deployment.Spec.Template.ObjectMeta.Annotations = map[string]string{}
+	}
+	deployment.Spec.Template.ObjectMeta.Annotations[kubernetes.AnnotationSecretHash] = hash
```
Comment (rynowak) [inline pkg/controller/reconciler/deployment_reconciler.go:569]: This makes sure we rollout new pods when the connections or connection data changes

File: pkg/controller/reconciler/deployment_reconciler_test.go [hunk 3]
```diff
@@ -145,14 +152,14 @@ func Test_DeploymentReconciler_ChangeEnvironmentAndApplication(t *testing.T) {
 
 	// Deployment will be waiting for container to complete deployment.
 	annotations := waitForStateUpdating(t, client, name)
-	require.Equal(t, "/planes/radius/local/resourceGroups/default-deployment-change-envapp", annotations.Status.Scope)
+	require.Equal(t, "/planes/radius/local/resourcegroups/default-deployment-change-envapp", annotations.Status.Scope)
 	require.Equal(t, "/planes/radius/local/resourceGroups/default/providers/Applications.Core/environments/default", annotations.Status.Environment)
-	require.Equal(t, "/planes/radius/local/resourceGroups/default-deployment-change-envapp/providers/Applications.Core/applications/deployment-change-envapp", annotations.Status.Application)
+	require.Equal(t, "/planes/radius/local/resourcegroups/default-deployment-change-envapp/providers/Applications.Core/applications/deployment-change-envapp", annotations.Status.Application)
```
Comment (ytimocin) [inline pkg/controller/reconciler/deployment_reconciler_test.go:157]: Should we make everything lower case?
```suggestion
	require.Equal(t, "/planes/radius/local/resourcegroups/default-deployment-change-envapp/providers/applications.core/applications/deployment-change-envapp", annotations.Status.Application)
```

**Review by lakshmimsft — APPROVED** [review]

File: pkg/kubernetes/secrets.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	for k := range secretData {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	hash := sha1.New()
+
+	for _, k := range keys {
+		// Using | as a delimiter
+		_, _ = hash.Write([]byte("|" + k + "|"))
```
Comment (ytimocin) [inline pkg/kubernetes/secrets.go:41]: Is this going to be something like `|key1|value1|key2|value2`?

**Review by ytimocin — APPROVED** [review]

File: pkg/controller/reconciler/deployment_reconciler_test.go (hunk 3 above)
Comment (ytimocin) [reply pkg/controller/reconciler/deployment_reconciler_test.go:157]: Just saw that you lower-cased the **resourceGroups**... You can ignore the comment above.

File: pkg/kubernetes/secrets.go (hunk 4 above)
Comment (rynowak) [reply pkg/kubernetes/secrets.go:41]: Yes. The format's not that important, just that it's unique given the input data.

**Pull request merged**